### PR TITLE
Expose include-hov/hot parameters and add exclude list option

### PR DESCRIPTION
**Pull request by Guardiola31337**
- Fixes https://github.com/mapbox/mapbox-java/issues/1293
- Adds `exclude` list option https://github.com/mapbox/mapbox-java/pull/1296#discussion_r714831962

Opening this as a `Draft` until Directions team deploys `include` params from their side in production. For now, we can test it in Staging using a `SNAPSHOT` from https://github.com/mapbox/mapbox-navigation-android

cc @browndp08 @ktatterso @kelseylivingston @zugaldia @jill-cardamon

### CODE REVIEW

File: services-directions-models/src/main/java/com/mapbox/api/directions/v5/DirectionsCriteria.java [hunk 1]
```diff
@@ -178,6 +178,21 @@
    */
   public static final String EXCLUDE_RESTRICTED = "restricted";
 
+  /**
+   * TBD.
```
Comment (Guardiola31337) [inline services-directions-models/src/main/java/com/mapbox/api/directions/v5/DirectionsCriteria.java:182]: @ktatterso could you share the docs?

File: services-directions-models/src/main/java/com/mapbox/api/directions/v5/DirectionsCriteria.java [hunk 2]
```diff
@@ -178,6 +178,21 @@
...
+   */
+  public static final String INCLUDE_HOV2 = "hov2";
+
+  /**
+   * TBD.
```
Comment (Guardiola31337) [inline services-directions-models/src/main/java/com/mapbox/api/directions/v5/DirectionsCriteria.java:187]: @ktatterso could you share the docs?

File: services-directions-models/src/main/java/com/mapbox/api/directions/v5/models/RouteOptions.java [hunk 3]
```diff
@@ -353,6 +353,26 @@ public static Builder builder() {
   @Nullable
   public abstract String exclude();
 
+  /**
+   * Include certain road types from routing. The default is to not include anything from the
+   * profile selected. The following include flags are available for each profile:
```
Comment (Guardiola31337) [inline services-directions-models/src/main/java/com/mapbox/api/directions/v5/models/RouteOptions.java:358]: @ktatterso is this accurate? If not, could you share the docs?

File: services-directions-models/src/main/java/com/mapbox/api/directions/v5/DirectionsCriteria.java [hunk 4]
```diff
@@ -178,6 +178,21 @@
...
+   */
+  public static final String INCLUDE_HOV3 = "hov3";
+
+  /**
+   * TBD.
```
Comment (Guardiola31337) [inline services-directions-models/src/main/java/com/mapbox/api/directions/v5/DirectionsCriteria.java:192]: @ktatterso could you share the docs?

**Comment by ktatterso** [comment]
@Guardiola31337 The doc changes for `include` are here:  https://github.com/mapbox/api-documentation/pull/1237

The api-valhalla changes for this parameter were only approved a couple hours ago.  They will be released on Monday (if things go according to schedule).  At which point I will merge these doc changes into master.

As you might imagine, we don't release documentation changes before the feature has been released.  Hence why I'm waiting.

Here are the "pretty" docs from that PR:
image

**Comment by zugaldia** [comment]
API support for this feature has landed, we're good to proceed merging this PR and issuing a release.

/cc: @browndp08

File: services-directions-models/src/main/java/com/mapbox/api/directions/v5/models/RouteOptions.java (hunk 3 above)
Comment (LukasPaczos) [reply services-directions-models/src/main/java/com/mapbox/api/directions/v5/models/RouteOptions.java:358]:
```suggestion
   * Include certain road types in routing. By default, none of the road types listed below are included. The following include flags are available for each profile:
```

File: services-directions-models/src/main/java/com/mapbox/api/directions/v5/models/RouteOptions.java [hunk 5]
```diff
@@ -1195,6 +1215,28 @@ public Builder annotationsList(@Nullable List<String> annotations) {
     @NonNull
     public abstract Builder exclude(@Nullable @DirectionsCriteria.ExcludeCriteria String exclude);
 
+    /**
+     * Include certain road types from routing. The default is to not include anything from the
+     * profile selected. The following include flags are available for each profile:
```
Comment (LukasPaczos) [inline services-directions-models/src/main/java/com/mapbox/api/directions/v5/models/RouteOptions.java:1220]: Same as above.

File: services-directions-models/src/main/java/com/mapbox/api/directions/v5/models/RouteOptions.java [hunk 6]
```diff
@@ -1195,6 +1215,28 @@ public Builder annotationsList(@Nullable List<String> annotations) {
...
+     * {@link DirectionsCriteria#PROFILE_WALKING}: No includes supported
+     * <p>
+     * {@link DirectionsCriteria#PROFILE_CYCLING}: No includes supported
+     *
+     * @param include a string matching one of the {@link DirectionsCriteria.IncludeCriteria}
+     *                inclusions
+     * @return this builder for chaining options together
+     */
+    @NonNull
+    public abstract Builder include(@Nullable @DirectionsCriteria.IncludeCriteria String include);
```
Comment (LukasPaczos) [inline services-directions-models/src/main/java/com/mapbox/api/directions/v5/models/RouteOptions.java:1238]: Can you include only one parameter, or can it be a list? If the latter, we should also expose a `List` method variant and test it.

Comment (Guardiola31337) [reply services-directions-models/src/main/java/com/mapbox/api/directions/v5/models/RouteOptions.java:1238]: Similar to `exclude` https://github.com/mapbox/mapbox-navigation-android/blob/d1218a77e305e04672754835df5fe22a4a68985a/libnavigation-base/src/main/java/com/mapbox/navigation/base/route/RouteExclusions.kt#L15-L21 should we do implement that for `include` from the SDK side for consistency? 🤔

Comment (LukasPaczos) [reply services-directions-models/src/main/java/com/mapbox/api/directions/v5/models/RouteOptions.java:1238]: Hmm, this should be part of the `mapbox-java` for both include and exclude, same as all the other list paramaters.

**Comment by Guardiola31337** [comment]
Addressed the feedback. This is ready for another round of eyes @mapbox/navigation-android

File: services-directions-models/src/test/java/com/mapbox/api/directions/v5/models/RouteOptionsTest.java [hunk 7]
```diff
@@ -387,7 +395,14 @@ private RouteOptions routeOptionsList() {
         add(0);
       }})
       .continueStraight(false)
-      .exclude(DirectionsCriteria.EXCLUDE_TOLL)
+      .excludeList(new ArrayList<String>() {{
+          add(DirectionsCriteria.EXCLUDE_TOLL);
+          add(DirectionsCriteria.EXCLUDE_TUNNEL);
```
Comment (abhishek1508) [inline services-directions-models/src/test/java/com/mapbox/api/directions/v5/models/RouteOptionsTest.java:400]: This test uses a `DirectionsCriteria.PROFILE_DRIVING_TRAFFIC` and excludes `TUNNEL`, but `TUNNEL` exclude is not mentioned in the documentation above. If `DRIVING` and `DRIVING_TRAFFIC` both support `TUNNEL` as exclude could you please add it to the documentation

**Review by abhishek1508 — COMMENTED** [review]
Could we add some more tests on what happens if we exclude or include some criteria for a profile that is not supported? For ex: a test for what happens if `Profile.Walking` and you add `Include.HOV2`

**Comment by Guardiola31337** [comment]
> Could we add some more tests on what happens if we exclude or include some criteria for a profile that is not supported?

I think those are developer misconfigurations that we shouldn't handle from the client side. We threw exceptions in the past but we decided it wasn't a good idea and let Directions API handle it. Actually if tomorrow Directions API wants to support one more profile we'd need to change it from our side and cut a new release. I'd vote for avoiding that.

File: services-directions-models/src/test/java/com/mapbox/api/directions/v5/models/RouteOptionsTest.java (hunk 7 above)
Comment (Guardiola31337) [reply services-directions-models/src/test/java/com/mapbox/api/directions/v5/models/RouteOptionsTest.java:400]: That is correct. That being said, Directions team is working on supporting the `DRIVING_TRAFFIC` profile in the short term https://github.com/mapbox/mapbox-navigation-android/issues/3844#issuecomment-743361069

In order to be more accurate, I can fix the tests to use `DRIVING` profile instead 👍

Comment (Guardiola31337) [reply services-directions-models/src/test/java/com/mapbox/api/directions/v5/models/RouteOptionsTest.java:400]: Addressed this.

**Comment by Guardiola31337** [comment]
Thanks for the feedback @abhishek1508 I've addressed it so this is ready again for another round of 👀

File: services-directions/src/main/java/com/mapbox/api/directions/v5/DirectionsService.java [hunk 8]
```diff
@@ -24,7 +24,7 @@
    * @param user                    {@link RouteOptions#user()}
    * @param profile                 {@link RouteOptions#profile()}
    * @param coordinates             {@link RouteOptions#coordinatesList()}
-   * @param accessToken             {@link RouteOptions#accessToken()}
+   * @param accessToken             the access token
```
Comment (kmadsen) [inline services-directions/src/main/java/com/mapbox/api/directions/v5/DirectionsService.java:27]: Why is this changing?

Comment (Guardiola31337) [reply services-directions/src/main/java/com/mapbox/api/directions/v5/DirectionsService.java:27]: Broken Javadocs, I was practicing the [The Boy Scout Rule](https://biratkirat.medium.com/step-8-the-boy-scout-rule-robert-c-martin-uncle-bob-9ac839778385) 🧹

Comment (kmadsen) [reply services-directions/src/main/java/com/mapbox/api/directions/v5/DirectionsService.java:27]: Oh never mind, RouteOptions does not have accessToken https://github.com/mapbox/mapbox-java/pull/1276

Comment (Guardiola31337) [reply services-directions/src/main/java/com/mapbox/api/directions/v5/DirectionsService.java:27]: Yeah, not anymore, some leftovers from https://github.com/mapbox/mapbox-java/pull/1276

File: services-directions-models/src/test/java/com/mapbox/api/directions/v5/models/RouteOptionsTest.java (hunk 7 above)
Comment (abhishek1508) [reply services-directions-models/src/test/java/com/mapbox/api/directions/v5/models/RouteOptionsTest.java:400]: Thanks for addressing the feedback, but I meant this:

```
A list of exclude. Exclude certain road types from routing.
     * The default is to not exclude anything from the profile selected.
     * The following exclude flags are available for each profile:
     * <p>
     * {@link DirectionsCriteria#PROFILE_DRIVING}: One of {@link DirectionsCriteria#EXCLUDE_TOLL},
     * {@link DirectionsCriteria#EXCLUDE_MOTORWAY}, or {@link DirectionsCriteria#EXCLUDE_FERRY}.
```
In the test you are using `EXCLUDE_TUNNEL` for `PROFILE_DRIVING`, but the documentation does not mention that it is supported. Could you add `{@link DirectionsCriteria#EXCLUDE_TUNNEL}` to the documenation?

Comment (Guardiola31337) [reply services-directions-models/src/test/java/com/mapbox/api/directions/v5/models/RouteOptionsTest.java:400]: Going to fix the tests and use `ferry` for example. `tunnel` is not supported based on https://docs.mapbox.com/api/navigation/directions/

| Profile                | Available excludes                               |
|------------------------|--------------------------------------------------|
| mapbox/driving         | One or more of `toll`, `motorway`, `ferry`, or `unpaved` |
| mapbox/driving-traffic | One or more of `toll`, `motorway`, `ferry`, or `unpaved` |
| mapbox/walking         | No excludes supported                            |
| mapbox/cycling         | `ferry`                                            |

Thanks for confirming @danpat

**Comment by Guardiola31337** [comment]
@kmadsen @abhishek1508 OK ready for a final round of 👀

**Review by abhishek1508 — APPROVED** [review]
Thanks for addressing the feedback @Guardiola31337

File: services-directions-models/src/main/java/com/mapbox/api/directions/v5/models/RouteOptions.java [hunk 9]
```diff
@@ -353,6 +353,72 @@ public static Builder builder() {
   @Nullable
   public abstract String exclude();
 
+  /**
+   * A list of exclude.
```
Comment (kmadsen) [inline services-directions-models/src/main/java/com/mapbox/api/directions/v5/models/RouteOptions.java:357]: Don't think this is needed. Redundant with the `@return` and there is already a description
```suggestion
```

File: services-directions-models/src/main/java/com/mapbox/api/directions/v5/models/RouteOptions.java [hunk 10]
```diff
@@ -353,6 +353,72 @@ public static Builder builder() {
...
+   * <p>
+   * {@link DirectionsCriteria#PROFILE_CYCLING}: No includes supported
+   *
+   * @return a string matching one of the {@link DirectionsCriteria.IncludeCriteria} inclusions
+   */
+  @Nullable
+  public abstract String include();
+
+  /**
+   * A list of include.
```
Comment (kmadsen) [inline services-directions-models/src/main/java/com/mapbox/api/directions/v5/models/RouteOptions.java:400]: Same as above, don't think this is needed.
```suggestion
```

File: services-directions-models/src/main/java/com/mapbox/api/directions/v5/models/RouteOptions.java [hunk 11]
```diff
@@ -1195,6 +1262,84 @@ public Builder annotationsList(@Nullable List<String> annotations) {
     @NonNull
     public abstract Builder exclude(@Nullable @DirectionsCriteria.ExcludeCriteria String exclude);
 
+    /**
+     * A list of exclude. Exclude certain road types from routing.
```
Comment (kmadsen) [inline services-directions-models/src/main/java/com/mapbox/api/directions/v5/models/RouteOptions.java:1266]:
```suggestion
     * Exclude certain road types from routing.
```

**Pull request merged**